Support for import.meta.url

Hi, I wish to target both esm and cjs. The package contains a server that serves static assets. Is there a way to refer to a relative folder to serve assets from in Typescript?

I tried to use __dirname but it isn't transpiled resulting in a runtime error in the esm build.
```
ReferenceError: __dirname is not defined in ES module scope
```

I tried to use import.meta.url but it fails to build with the following error:
```
src/index.ts(1,24): error TS1343: The 'import.meta' meta-property is only allowed
when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
      
       ⨯ RollupError: Failed to compile. Check the logs above.
```

It seems like the values in tsconfig.json don't affect the error.

--- a/package.json
+++ b/package.json
@@ -48,6 +48,7 @@
   "license": "MIT",
   "dependencies": {
     "@rollup/plugin-commonjs": "^25.0.4",
+    "@rollup/plugin-esm-shim": "^0.1.5",
     "@rollup/plugin-json": "^6.0.0",
     "@rollup/plugin-node-resolve": "^15.2.1",
     "@rollup/plugin-replace": "^5.0.2",

--- a/src/build-config.ts
+++ b/src/build-config.ts
@@ -17,6 +17,7 @@ import commonjs from '@rollup/plugin-commonjs'
 import json from '@rollup/plugin-json'
 import { nodeResolve } from '@rollup/plugin-node-resolve'
 import replace from '@rollup/plugin-replace'
+import esmShim from '@rollup/plugin-esm-shim'
 import { sizeCollector } from './plugins/size-plugin'
 import { inlineCss } from './plugins/inline-css'
 import swcPreserveDirectivePlugin from 'rollup-swc-preserve-directives'
@@ -192,6 +193,7 @@ async function buildInputConfig(
             tsconfig: tsConfigPath,
             ...swcOptions,
           }),
+          esmShim()
         ]
   ).filter(isNotNull<Plugin>)
 

--- a/test/integration.test.ts
+++ b/test/integration.test.ts
@@ -316,10 +316,23 @@ const testCases: {
     name: 'esm-shims',
     args: [],
     async expected(dir) {
-      expect(await fs.readFile(join(dir, './dist/index.mjs'), 'utf-8'))
-        .toContain(`const __filename = cjsUrl.fileURLToPath(import.meta.url);
-const __dirname = cjsPath.dirname(__filename);
-const require = cjsModule.createRequire(import.meta.url);`)
+      const shimsCode = [
+        'const __filename = cjsUrl.fileURLToPath(import.meta.url)',
+        'const __dirname = cjsPath.dirname(__filename)',
+        'const require = cjsModule.createRequire(import.meta.url)',
+      ]
+      const esmOutput = await fs.readFile(join(dir, './dist/index.mjs'), 'utf-8')
+      const cjsOutput = await fs.readFile(join(dir, './dist/index.cjs'), 'utf-8')
+      expect(
+        shimsCode.every((code) => esmOutput.includes(code)),
+      ).toBe(true)
+      expect(
+        shimsCode.map((code) => cjsOutput.includes(code)),
+      ).toEqual([false, false, false])
+      // for import.meta.url, should use pathToFileURL + URL polyfill
+      expect(cjsOutput).toContain('pathToFileURL')
+      expect(cjsOutput).toContain('new URL')
+      expect(cjsOutput).not.toContain('import.meta.url')
     },
   },
 ]

--- a/test/integration.test.ts
+++ b/test/integration.test.ts
@@ -228,7 +228,10 @@ const testCases: {
     name: 'ts-incremental',
     args: [],
     async expected(dir) {
-      const distFiles = ['./dist/index.js', './dist/index.d.ts']
+      const distFiles = [
+        './dist/index.js',
+        './dist/index.d.ts',
+      ]
 
       for (const f of distFiles) {
         expect(await existsFile(join(dir, f))).toBe(true)

--- a/test/integration.test.ts
+++ b/test/integration.test.ts
@@ -228,10 +228,7 @@ const testCases: {
     name: 'ts-incremental',
     args: [],
     async expected(dir) {
-      const distFiles = [
-        './dist/index.js',
-        './dist/index.d.ts',
-      ]
+      const distFiles = ['./dist/index.js', './dist/index.d.ts']
 
       for (const f of distFiles) {
         expect(await existsFile(join(dir, f))).toBe(true)
@@ -310,6 +307,16 @@ const testCases: {
       expect(stderr).toContain(
         'Cannot export main field with .cjs extension in ESM package, only .mjs and .js extensions are allowed',
       )
+    },
+  },
+  {
+    name: 'esm-shims',
+    args: [],
+    async expected(dir) {
+      expect(await fs.readFile(join(dir, './dist/index.mjs'), 'utf-8'))
+        .toContain(`const __filename = cjsUrl.fileURLToPath(import.meta.url);
+const __dirname = cjsPath.dirname(__filename);
+const require = cjsModule.createRequire(import.meta.url);`)
     },
   },
 ]

--- a/test/integration/esm-shims/package.json
+++ b/test/integration/esm-shims/package.json
@@ -1,7 +1,8 @@
 {
   "exports": {
     ".": {
-      "import": "./dist/index.mjs"
+      "import": "./dist/index.mjs",
+      "require": "./dist/index.cjs"
     }
   }
 }

--- a/test/integration/esm-shims/package.json
+++ b/test/integration/esm-shims/package.json
@@ -1,0 +1,7 @@
+{
+  "exports": {
+    ".": {
+      "import": "./dist/index.mjs"
+    }
+  }
+}

--- a/test/integration/esm-shims/src/index.js
+++ b/test/integration/esm-shims/src/index.js
@@ -1,0 +1,3 @@
+console.log(__dirname)
+console.log(__filename)
+console.log(import.meta.url)

